Python3.14 breaks pre-commit: AttributeError: module 'ast' has no attribute 'NameConstant'

It seems that Python3.14 is breaking flake8 in the pre-commit due to a removed/deprecated `ast.NameConstant`. 

There hasn't been any changes to the pre-commit config since the last time a PR landed in `main`. But back then the [Python version was still at 3.13](https://github.com/psss/did/actions/runs/18776128104/job/53571057827#step:9:5).

Here's an output from the broken `flake8`:

```
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ~~~~^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8/checker.py", line 82, in _mp_run
    ).run_checks()
      ~~~~~~~~~~^^
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8/checker.py", line 526, in run_checks
    self.run_ast_checks()
    ~~~~~~~~~~~~~~~~~~~^^
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8/checker.py", line 428, in run_ast_checks
    for line_number, offset, text, _ in runner:
                                        ^^^^^^
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8_plugin_utils/plugin.py", line 71, in run
    visitor.visit(self._tree)
    ~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 506, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 514, in generic_visit
    self.visit(item)
    ~~~~~~~~~~^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 506, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 514, in generic_visit
    self.visit(item)
    ~~~~~~~~~~^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 506, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 514, in generic_visit
    self.visit(item)
    ~~~~~~~~~~^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 506, in visit
    return visitor(node)
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 514, in generic_visit
    self.visit(item)
    ~~~~~~~~~~^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/ast.py", line 506, in visit
    return visitor(node)
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8_pytest_style/visitors/fail.py", line 30, in visit_Assert
    if is_falsy_constant(node.test):
       ~~~~~~~~~~~~~~~~~^^^^^^^^^^^
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8_pytest_style/utils.py", line 265, in is_falsy_constant
    if is_none(node) or is_false(node):
       ~~~~~~~^^^^^^
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8_plugin_utils/utils/constants.py", line 9, in is_none
    return _is(node, None)
  File "/home/runner/.cache/pre-commit/repoo7eb0p9i/py_env-python3.14/lib/python3.14/site-packages/flake8_plugin_utils/utils/constants.py", line 5, in _is
    return isinstance(node, ast.NameConstant) and node.value is value
                            ^^^^^^^^^^^^^^^^
AttributeError: module 'ast' has no attribute 'NameConstant'
"""
did/plugins/public_inbox.py:27:0: E0401: Unable to import 'tenacity' (import-error)
************* Module did.plugins.redmine
did/plugins/redmine.py:22:0: E0401: Unable to import 'dateutil' (import-error)
did/plugins/redmine.py:23:0: E0401: Unable to import 'feedparser' (import-error)
************* Module did.plugins.rt
did/plugins/rt.py:16:0: E0401: Unable to import 'gssapi' (import-error)
************* Module did.plugins.sentry
did/plugins/sentry.py:25:0: E0401: Unable to import 'dateutil' (import-error)
did/plugins/sentry.py:26:0: E0401: Unable to import 'requests' (import-error)
************* Module did.stats
did/stats.py:9:0: E0611: No name 'ThreadPoolExecutor' in module 'concurrent.futures' (no-name-in-module)
************* Module setup
setup.py:5:0: E0401: Unable to import 'setuptools' (import-error)
************* Module tests.plugins.test_bodhi
tests/plugins/test_bodhi.py:8:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_bugzilla
tests/plugins/test_bugzilla.py:6:0: E0401: Unable to import 'pytest' (import-error)
tests/plugins/test_bugzilla.py:7:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_confluence
tests/plugins/test_confluence.py:6:0: E0401: Unable to import 'pytest' (import-error)
tests/plugins/test_confluence.py:7:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_gerrit
tests/plugins/test_gerrit.py:4:0: E0401: Unable to import 'pytest' (import-error)
************* Module tests.plugins.test_git
tests/plugins/test_git.py:7:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_github
tests/plugins/test_github.py:9:0: E0401: Unable to import 'pytest' (import-error)
tests/plugins/test_github.py:10:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_gitlab
tests/plugins/test_gitlab.py:7:0: E0401: Unable to import 'pytest' (import-error)
tests/plugins/test_gitlab.py:8:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_google
tests/plugins/test_google.py:8:0: E0401: Unable to import 'pytest' (import-error)
************* Module tests.plugins.test_hyperkitty
tests/plugins/test_hyperkitty.py:6:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_jira
tests/plugins/test_jira.py:9:0: E0401: Unable to import 'pytest' (import-error)
tests/plugins/test_jira.py:10:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_koji
tests/plugins/test_koji.py:8:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_pagure
tests/plugins/test_pagure.py:10:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_phabricator
tests/plugins/test_phabricator.py:8:0: E0401: Unable to import 'pytest' (import-error)
************* Module tests.plugins.test_sentry
tests/plugins/test_sentry.py:6:0: E0401: Unable to import 'pytest' (import-error)
tests/plugins/test_sentry.py:7:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_trac
tests/plugins/test_trac.py:6:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_trello
tests/plugins/test_trello.py:8:0: E0401: Unable to import 'pytest' (import-error)
tests/plugins/test_trello.py:9:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.plugins.test_zammad
tests/plugins/test_zammad.py:6:0: E0401: Unable to import '_pytest.logging' (import-error)
************* Module tests.test_base
tests/test_base.py:13:0: E0401: Unable to import 'pytest' (import-error)
************* Module tests.test_cli
tests/test_cli.py:7:0: E0401: Unable to import 'pytest' (import-error)
************* Module tests.test_stats
tests/test_stats.py:5:0: E0401: Unable to import 'pytest' (import-error)
************* Module tests.test_utils
tests/test_utils.py:8:0: E0401: Unable to import 'pytest' (import-error)
tests/test_utils.py:9:0: E0401: Unable to import '_pytest.logging' (import-error)
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -15,7 +15,7 @@ repos:
           - --max-line-length=88
 
   - repo: https://github.com/PyCQA/isort
-    rev: "6.0.1"
+    rev: "7.0.0"
     hooks:
       - id: isort
 
@@ -25,7 +25,8 @@ repos:
       - id: flake8
         additional_dependencies:
           [
-            flake8-pytest-style == 2.1.0,
+            # https://pypi.org/project/flake8-pytest-style/
+            flake8-pytest-style == 2.2.0,
           ]
         args:
           - --config
@@ -45,7 +46,7 @@ repos:
           )$
 
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: "v5.0.0"
+    rev: "v6.0.0"
     hooks:
       - id: end-of-file-fixer
       - id: mixed-line-ending
@@ -56,7 +57,7 @@ repos:
         exclude: disable_test_trac.py
 
   - repo: https://github.com/pylint-dev/pylint
-    rev: "v3.3.7"
+    rev: "v4.0.2"
     hooks:
       - id: pylint
         args:
@@ -82,7 +83,7 @@ repos:
           ]
 
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: 'v1.17.0'
+    rev: 'v1.18.2'
     hooks:
       - id: mypy
         exclude: examples/.*|bin/did

--- a/.pylintrc
+++ b/.pylintrc
@@ -1,6 +1,6 @@
 [MAIN]
 ignore-paths=^bin/.*$
-init-hook='import sys; import site; sys.path.extend(site.getsitepackages()); sys.path.append("./.venv/lib64/python3.13/site-packages"); sys.path.append("/usr/lib/python3.13/site-packages")'
+init-hook='import sys; import site; sys.path.extend(site.getsitepackages()); sys.path.append("./.venv/lib64/python3.14/site-packages"); sys.path.append("/usr/lib/python3.13/site-packages")'
 
 [MESSAGES]
 disable=duplicate-code,missing-function-docstring,missing-module-docstring,missing-class-docstring,fixme,too-many-arguments,too-many-instance-attributes

--- a/did/base.py
+++ b/did/base.py
@@ -1,6 +1,5 @@
 """ Config, Date, User and Exceptions """
 
-import codecs
 import configparser
 import contextlib
 import datetime
@@ -129,7 +128,7 @@ class Config():
         # Parse the config from file
         try:
             log.info("Inspecting config file '%s'.", path)
-            with codecs.open(path, "r", "utf8") as config_file:
+            with open(path, "r", encoding="utf8") as config_file:
                 Config.parser.read_file(config_file)
         except IOError as error:
             log.debug(error)

--- a/setup.py
+++ b/setup.py
@@ -34,7 +34,8 @@ extras_require = {
     'redmine': ['feedparser'],
     'nitrate': ['nitrate'],
     'rt': ['gssapi'],
-    'tests': ['pytest', 'pytest-xdist', 'pytest-cov', 'python-coveralls', 'pre-commit'],
+    'tests': ['pytest', 'pytest-xdist', 'pytest-cov', 'python-coveralls', 'pre-commit',
+              'setuptools'],
     'mypy': ['types-setuptools', 'types-python-dateutil', 'lxml',
              'types-requests', 'types-urllib3', 'types-httplib2'],
     }
@@ -75,6 +76,7 @@ setup(
         'Programming Language :: Python :: 3.11',
         'Programming Language :: Python :: 3.12',
         'Programming Language :: Python :: 3.13',
+        'Programming Language :: Python :: 3.14',
         'Topic :: Office/Business',
         'Topic :: Utilities',
         ],

